draft-ietf-quic-transport.md

```diff
@@ -1194,9 +1194,46 @@ A client MAY attempt 0-RTT after receiving a Version Negotiation packet.  A
 client that sends additional 0-RTT packets MUST NOT reset the packet number to 0
 as a result, see {{retry-0rtt-pn}}.
 
-Version negotiation packets have no cryptographic protection. The result of the
-negotiation MUST be revalidated as part of the cryptographic handshake (see
-{{version-validation}}).
+The format of the packet that a client sends might be different in the new
+version.  In this case, the client generates a new packet that conforms to the
+selected version.  Though different versions might convey information about
+versions differently, a client MUST NOT change either the list of versions it
+claims to support.  Similarly, a client MUST NOT change what it reports as the
+version it first attempted to use.
+
+Version negotiation packets have no cryptographic protection.  Keeping
+version-related information consistent between versions is critical for
+validation of the choice of version (see {{version-validation}}).
+
+
+## Compatible Version Upgrade {#version-upgrade}
+
+The first packet sent by a client uses a QUIC version selected by the client.
+In this version of QUIC, this packet also includes transport parameters.  Those
+transport parameters include a list of QUIC versions the client supports.
+
+A server that understands the version selected by the client can extract the
+list of QUIC versions and select an alternative version from the list of
+supported versions.  A server MAY choose a compatible version from that list and
+continue the handshake with that version.  The server sends its first packet as
+though it was continuing with the version it selects.
+
+A QUIC version is compatible with this version if the cryptographic handshake
+message sent in the first packet can be used in both versions.  A compatible
+version is also able to identifying and acknowledge the first packet sent by the
+client in some fashion.  Other QUIC versions might have different constraints in
+determining what is compatible.  In order to facilitate this process, new QUIC
+versions could define a process for transforming the first packet from other
+compatible versions into the equivalent packet in the new version.
+
+Upgrading in this manner allows a server to upgrade without incurring the round
+trip imposed by sending a Version Negotiation packet.  It also allows clients to
+send their first packet using a widely deployed version, without the risk of
```

Comment:
There’s no incentive to use a widely deployed version any more. A client can use the highest (compatible) version, and the server would just downgrade it to the highest version it supports.

Reply:
The server might not know that there is a newer compatible version - compatible is both a spec thing and an implementation thing.

Reply:
Thinking of versions as being ordered is helpful, but might be constraining.  The point is that the client will use the most widely deployed to start with and accept that the server is able to choose a compatible version.  The client *hopes* that the server will choose a version that it prefers more, but leaves that to the server to decide.

In the case of monotonically increasing versions where A >> B >> C, this might seem obvious, but it's still possible that the client will advertise B and get an apparent downgrade to C.  That's a property of the design that I don't think we'll see that often, but it's nonetheless theoretically possible.

But as @kazuho says, the expected case is a client advertising vN, then being opportunistically upgraded to vN+1.  Like today, you can rely on TLS 1.0 being available widely, but you might prefer TLS 1.3.

Reply:
The TLS comparison is somewhat inapt, since in TLS you actually do offer your highest (most preferred) version and let the server select a compatible, less-preferred version if it doesn't support that one.  QUIC can't do that because the contents of the Initial aren't invariant.